src/mesh/boundary_nonuniform.cxx

```diff
@@ -0,0 +1,1285 @@
+      by = -1;
+    } else if (by < 0) {
+      stag = -1;
+    } else {
+      stag = 1;
+    }
+  }
+  int istart = (stag == -1) ? -1 : 0;
+
+  for (; !bndry->isDone(); bndry->next1d()) {
+    if (fg) {
+      for (int zk = 0; zk < mesh->LocalNz; zk++) {
+        // Calculate the X and Y normalised values half-way between the guard cell and
+        // grid cell
+        BoutReal xnorm = 0.5
+                         * (mesh->GlobalX(bndry->x)          // In the guard cell
+                            + mesh->GlobalX(bndry->x - bx)); // the grid cell
+
+        BoutReal ynorm = 0.5
+                         * (mesh->GlobalY(bndry->y)          // In the guard cell
+                            + mesh->GlobalY(bndry->y - by)); // the grid cell
+
+        vals[zk] = fg->generate(xnorm, TWOPI * ynorm, TWOPI * zk / (mesh->LocalNz), t);
+      }
+    }
+    BoutReal fac0;
+    BoutReal x0;
+    BoutReal fac1;
+    BoutReal x1;
+
+    const Field2D& dy =
+        bndry->by != 0 ? mesh->getCoordinates()->dy : mesh->getCoordinates()->dx;
+    Indices i1{bndry->x - 1 * bndry->bx, bndry->y - 1 * bndry->by, 0};
+    BoutReal t;
+    if (stag == 0) {
+      x0 = 0;
+      BoutReal st = 0;
+      t = dy[i1];
+      x1 = st + t / 2;
+      st += t;
+    } else {
+      x0 = 0; // dy(bndry->x, bndry->y) / 2;
+      x1 = x0 + dy[i1];
+    }
+    if (stag == -1) {
+      i1 = {bndry->x - 2 * bndry->bx, bndry->y - 2 * bndry->by, 0};
+    }
+    for (int i = istart; i < bndry->width; i++) {
+      Indices ic{bndry->x + i * bndry->bx, bndry->y + i * bndry->by, 0};
+      if (stag == 0) {
+        t = dy[ic] / 2;
+        x0 += t;
+        x1 += t;
+        // printf("%+2d: %d %d %g %g %g %g\n", stag, ic.x, ic.y, x0, x1, x2, x3);
+        calc_interp_to_stencil(x0, x1, fac0, fac1);
+        x0 += t;
+        x1 += t;
+      } else {
+        t = dy[ic];
+        if (stag == -1 && i != -1) {
+          x0 += t;
+          x1 += t;
+        }
+        calc_interp_to_stencil(x0, x1, fac0, fac1);
+        if (stag == 1) {
+          x0 += t;
+          x1 += t;
+        }
+      }
+      for (ic.z = 0; ic.z < mesh->LocalNz; ic.z++) {
+        i1.z = ic.z;
+        val = (fg) ? vals[ic.z] : 0.0;
+        t = fac0 * val + fac1 * f[i1];
+
+        f[ic] = t;
+      }
+    }
+  }
+}
+
+BoundaryOp* BoundaryDirichletNonUniform_O2::clone(BoundaryRegion* region,
+                                                  const std::list<std::string>& args) {
+  // verifyNumPoints(region, 3);
+
+  std::shared_ptr<FieldGenerator> newgen;
+  if (!args.empty()) {
+    // First argument should be an expression
+    newgen = FieldFactory::get()->parse(args.front());
+  }
+  return new BoundaryDirichletNonUniform_O2(region, newgen);
+}
+
+void BoundaryDirichletNonUniform_O2::calc_interp_to_stencil(BoutReal x0, BoutReal x1,
+                                                            BoutReal& fac0,
+                                                            BoutReal& fac1) const {
```

Comment:
If `fac0`, `fac1` were in a struct, they could be a return value rather than out parameters